### PR TITLE
Always return a new tensor from nn.functional.pad

### LINKED ISSUE #31734: functional.pad returns original tensor when padding size is 0
## To Reproduce

```python
import torch
from torch.nn import functional as F
x = torch.ones(3, 10)

y1 = F.pad(x, [1, 1, 1, 1])
y2 = F.pad(x, [0, 0, 0, 0])
y3 = x + 0

y1.unsqueeze_(0)
print(x.shape)  # 3, 10

y3.unsqueeze_(0)
print(x.shape)  # 3, 10

y2.unsqueeze_(0)
print(x.shape)  # 1, 3, 10
print(x is y2)  # True

y1.requires_grad = True
print(x.requires_grad)  # False
y3.requires_grad = True
print(x.requires_grad)  # False
y2.requires_grad = True
print(x.requires_grad)  # True

```

## Expected behavior

I expect y2 to be a different tensor from x, although I understand this may be arguable.

In the current situation, the semantic of `F.pad` is different depending the padded #pixels: its output may be the same input tensor sometimes. And modifying the output tensor will __sometimes__ modify the input tensor - potentially lead to unnoticed dangerous consequences. 
As a related note, `x + 0` and `x[:, :]` is not the same tensor as `x` and I think doing so would be a bad idea.

To avoid such dangerous consequences, currently one will need to handle padding_size=0 manually.

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #31734

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit b06bdcd5:


**None of the build failures appear to be your fault**.

* **9/9** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=b06bdcd5faf95ae1880957036998db6dec9325f1).

### :snowflake: 9 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289944) pytorch_linux_xenial_py3_clang5_mobile_build (1/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87779050) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87779050)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289951) pytorch_libtorch_linux_xenial_cuda9_cudnn7_py3_build (2/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87779051) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87779051)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289957) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_mobile_code_analysis (3/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87779201) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87779201)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289964) pytorch_linux_xenial_py3_clang5_asan_build (4/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87779958) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87779958)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-asan:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289961) pytorch_linux_xenial_py2_7_9_build (5/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87780477) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87780477)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py2.7.9:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py2.7.9:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py2.7.9:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289950) pytorch_xla_linux_xenial_py3_6_clang7_build (6/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87780512) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87780512)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-clang7:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-clang7:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-clang7:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289956) pytorch_linux_xenial_py3_6_gcc5_4_build (7/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87780557) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87780557)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-gcc5.4:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-gcc5.4:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-gcc5.4:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289954) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_x86_32_build (8/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87780559) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87780559)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4289953) pytorch_linux_xenial_cuda9_cudnn7_py3_build (9/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87780631) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87780631)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 not found</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda9-cudnn7-py3:a8006f9a-272d-4478-b137-d121c6f05c83 not found 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Review by ezyang — APPROVED** [review]
